pkg/internal/controller/controller.go

```diff
@@ -144,6 +220,39 @@ func (c *Controller[request]) NeedLeaderElection() bool {
 	return *c.LeaderElected
 }
 
+// Warmup implements the manager.WarmupRunnable interface.
+func (c *Controller[request]) Warmup(ctx context.Context) error {
+	if c.EnableWarmup == nil || !*c.EnableWarmup {
+		return nil
+	}
+
+	// Hold the lock to avoid concurrent access to c.startWatches with Start() when calling
+	// startEventSources
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	err := c.startEventSources(ctx)
+	c.warmupResultChan <- err
+
+	return err
+}
+
+// WaitForWarmupComplete returns true if warmup has completed without error, and false if there was
+// an error during warmup. If context is cancelled, it returns true.
+func (c *Controller[request]) WaitForWarmupComplete(ctx context.Context) bool {
```

Comment:
THere is no need for this anymore now I think?

Reply:
What do you mean with "this" the whole func?

Reply:
Discussed. The idea is to get rid of the entire func. Let's resolve to dedup with: https://github.com/kubernetes-sigs/controller-runtime/pull/3192#discussion_r2098142231